### PR TITLE
Uri component

### PR DESCRIPTION
**Pull request by alexandre-daubois**
| Q             | A
| ------------- | ---
| Branch?       | 7.1
| Bug fix?      | no
| New feature?  | yes
| Deprecations? | no
| Issues        | -
| License       | MIT

Follows [our specification](https://github.com/alexandre-daubois/uri-component-specs)

### CODE REVIEW

File: src/Symfony/Component/Uri/CHANGELOG.md [hunk 1]
```diff
@@ -0,0 +1,7 @@
+CHANGELOG
+=========
+
+7.1
```
Comment (Nyholm) [inline src/Symfony/Component/Uri/CHANGELOG.md]:
```suggestion
7.2
```

File: src/Symfony/Component/Uri/QueryString.php [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+
+/**
+ * @author Alexandre Daubois <alex.daubois@gmail.com>
+ */
+final class QueryString implements \Stringable
+{
+    /**
+     * @var array<string, string|string[]>
+     */
+    private array $hashMap = [];
```
Comment (Nyholm) [inline src/Symfony/Component/Uri/QueryString.php]: Let's give this a better name. Like $parameters?

File: src/Symfony/Component/Uri/Uri.php [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+     * Uri::resolve('http://example.com/foo', 'bar'); // http://example.com/foo/bar
+     * Uri::resolve('http://example.com/foo/', 'bar'); // http://example.com/foo/bar
+     * Uri::resolve('http://example.com/foo/', '../bar'); // http://example.com/bar
+     * Uri::resolve('http://example.com/foo/', '../../bar'); // http://example.com/bar
+     * Uri::resolve('http://example.com/foo/', '/bar'); // http://example.com/bar
+     * Uri::resolve('http://example.com/foo/', 'http://example.org/bar'); // http://example.org/bar
+     */
+    public static function resolve(Uri|string $baseUri, string $relativeUri): string
+    {
+        $baseUri = $baseUri instanceof Uri ? $baseUri : self::parse($baseUri);
```
Comment (Nyholm) [inline src/Symfony/Component/Uri/Uri.php]: Micro optimisation: Move this down to line 99.

**Review by Nyholm — APPROVED** [review]
Awesome. Looks good.

I've just reviewed the code. I will review the code + spec later.

File: src/Symfony/Component/Uri/QueryString.php (hunk 2 above)
Comment (alexandre-daubois) [reply src/Symfony/Component/Uri/QueryString.php]: Yes!

File: src/Symfony/Component/Uri/Uri.php (hunk 3 above)
Comment (alexandre-daubois) [reply src/Symfony/Component/Uri/Uri.php]: Good idea

**Comment by alexandre-daubois** [comment]
Thank you Tobias for your review!

I just pushed a few improvements in https://github.com/alexandre-daubois/symfony/pull/11/commits/7bcb8b61972a901f7e9959b569f4ea19553c7537:
- Fixed user and password reencoding when using `Uri::__toString()`
- Added a few more tests
- Implemented fragment text directives [as we specified here](https://github.com/alexandre-daubois/uri-component-specs?tab=readme-ov-file#fragment-text-directives), cc @javiereguiluz as you proposed this feature 🙂 

All the most important features that are "pretty sure" to make it to this hypothetic new component are now implemented. The other ones we mentioned could wait later iterations.

File: src/Symfony/Component/Uri/.gitattributes [hunk 4]
```diff
@@ -0,0 +1,4 @@
+Tests export-ignore
+/phpunit.xml.dist export-ignore
+/.gitattributes export-ignore
+/.gitignore export-ignore
```
Comment (OskarStark) [inline src/Symfony/Component/Uri/.gitattributes]:
```suggestion
/.git* export-ignore
```
As Nicolas changed that recently

File: src/Symfony/Component/Uri/CHANGELOG.md [hunk 5]
```diff
@@ -0,0 +1,7 @@
+CHANGELOG
+=========
+
+7.2
+---
+
+* Add the component
```
Comment (OskarStark) [inline src/Symfony/Component/Uri/CHANGELOG.md]:
```suggestion
 * Add the component as experimental
```

File: src/Symfony/Component/Uri/README.md [hunk 6]
```diff
@@ -0,0 +1,21 @@
+TypeInfo Component
```
Comment (OskarStark) [inline src/Symfony/Component/Uri/README.md]:
```suggestion
Uri Component
```

File: src/Symfony/Component/Uri/README.md [hunk 7]
```diff
@@ -0,0 +1,21 @@
...
+composer require symfony/uri
+```
+
+Resources
+---------
+
+* [Contributing](https://symfony.com/doc/current/contributing/index.html)
+* [Report issues](https://github.com/symfony/symfony/issues) and
+  [send Pull Requests](https://github.com/symfony/symfony/pulls)
+  in the [main Symfony repository](https://github.com/symfony/symfony)
```
Comment (OskarStark) [inline src/Symfony/Component/Uri/README.md]:
```suggestion
 * [Contributing](https://symfony.com/doc/current/contributing/index.html)
 * [Report issues](https://github.com/symfony/symfony/issues) and
   [send Pull Requests](https://github.com/symfony/symfony/pulls)
   in the [main Symfony repository](https://github.com/symfony/symfony)
```

File: src/Symfony/Component/Uri/Tests/FragmentTextDirectiveTest.php [hunk 8]
```diff
@@ -0,0 +1,53 @@
...
+    {
+        $fragmentTextDirective = new FragmentTextDirective('st&rt', 'e,nd', 'prefix-', '-&suffix');
+
+        $this->assertSame(':~:text=prefix%2D-,st%26rt,e%2Cnd,-%2D%26suffix', (string) $fragmentTextDirective);
+    }
+
+    public static function provideValidFragmentTextDirectives(): iterable
+    {
+        yield [new FragmentTextDirective('start'), ':~:text=start'];
+
```
Comment (OskarStark) [inline src/Symfony/Component/Uri/Tests/FragmentTextDirectiveTest.php]: No need for eh blank lines

File: src/Symfony/Component/Uri/Tests/UriTest.php [hunk 9]
```diff
@@ -0,0 +1,244 @@
...
+    {
+        $resolvedUri = Uri::resolve($baseUri, $relativeUri);
+
+        $this->assertSame($expectedUri, $resolvedUri);
+    }
+
+    public static function provideResolveUri(): iterable
+    {
+        yield "Single-level absolute path" => ['https://example.com/foo', '/bar', 'https://example.com/bar'];
+
```
Comment (OskarStark) [inline src/Symfony/Component/Uri/Tests/UriTest.php]: no need for the blank lines

File: src/Symfony/Component/Uri/composer.json [hunk 10]
```diff
@@ -0,0 +1,32 @@
+{
+    "name": "symfony/uri",
+    "type": "library",
+    "description": "Parses URLs",
+    "keywords": [
+        "uri",
+        "url",
+        "symfony"
+    ],
```
Comment (OskarStark) [inline src/Symfony/Component/Uri/composer.json]: AFAIK we keep it on one line

**Review by OskarStark — COMMENTED** [review]
The new component should be introduced as experimental

File: src/Symfony/Component/Uri/README.md (hunk 6 above)
Comment (alexandre-daubois) [reply src/Symfony/Component/Uri/README.md]: Oops 🙃 Thanks

**Comment by alexandre-daubois** [comment]
Addressed your comments. Thanks Oskar!

**Comment by alexandre-daubois** [comment]
I pushed a critical missing feature, which is the support for **hash maps in query strings** (just like `parse_str()` does). It's been highlighted while leveraging the Uri component in a real world use case. This is now solved in https://github.com/alexandre-daubois/symfony/pull/11/commits/0e22428d90fc0e9972134abe1f2642829d94e0f2.

**IDN hosts support** has been added in https://github.com/alexandre-daubois/symfony/pull/11/commits/205d4c460787b7f0e3f63b79d18b1cae7d230273.

The real world use case can be found here, where I leveraged the component in the Redis cache adapter for demo purpose: https://github.com/alexandre-daubois/symfony/pull/12.

It's starting to look pretty good!

**Comment by alexandre-daubois** [comment]
Closing in favor of https://github.com/symfony/symfony/pull/57192. Thank you everyone for your time and review on this draft!

**Pull request closed without merging**